### PR TITLE
Improved FileResource and DirectoryResource

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | yes? |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #15946 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

File: src/Symfony/Component/Config/Resource/FileResource.php [hunk 1]
```diff
@@ -29,10 +29,16 @@ class FileResource implements SelfCheckingResourceInterface, \Serializable
...
      *
      * @param string $resource The file path to the resource
+     *
+     * @throws \InvalidArgumentException
      */
     public function __construct($resource)
     {
         $this->resource = realpath($resource);
+
+        if (!file_exists($this->resource)) {
```
Comment (Tobion) [inline src/Symfony/Component/Config/Resource/FileResource.php]: realpath already returns false if it doesn't exist. so this is useless.

File: src/Symfony/Component/Config/Resource/DirectoryResource.php [hunk 2]
```diff
@@ -26,11 +26,17 @@ class DirectoryResource implements SelfCheckingResourceInterface, \Serializable
...
+     *
+     * @throws \InvalidArgumentException
      */
     public function __construct($resource, $pattern = null)
     {
-        $this->resource = $resource;
+        $this->resource = realpath($resource);
         $this->pattern = $pattern;
+
+        if (!is_dir($this->resource)) {
```
Comment (Tobion) [inline src/Symfony/Component/Config/Resource/DirectoryResource.php]: need to check first if $this->resource is not false

**Comment by Tobion** [comment]
1. phpdoc of FiileResource::getResource is not correct anymore
2. string cast in __toString doesnt make sense anymore
3. the check in isFresh `false === $this->resource` does not make sense

Status: Needs Work

**Comment by javiereguiluz** [comment]
@Tobion I've fixed everything you said. Thanks for the review!

File: src/Symfony/Component/Config/Resource/FileResource.php [hunk 3]
```diff
@@ -29,10 +29,16 @@ class FileResource implements SelfCheckingResourceInterface, \Serializable
...
      * @param string $resource The file path to the resource
+     *
+     * @throws \InvalidArgumentException
      */
     public function __construct($resource)
     {
         $this->resource = realpath($resource);
+
+        if (false === $this->resource) {
+            throw new \InvalidArgumentException(sprintf('The "%s" file does not exist.', $resource));
```
Comment (Tobion) [inline src/Symfony/Component/Config/Resource/FileResource.php]: usually you write it the other way round: The file "%s" does not exist. (same for directory)

Comment (javiereguiluz) [inline src/Symfony/Component/Config/Resource/FileResource.php]: You are right. Changed. Thanks.

**Comment by Tobion** [comment]
FileResource::isFresh is not safe against race conditions (when it's deleted between file_exists and filemtime). So it probably need error supression instead with `@filemtime`.

**Comment by javiereguiluz** [comment]
Should I replace this:

```php
public function isFresh($timestamp)
{
    if (!file_exists($this->resource)) {
        return false;
    }

    return filemtime($this->resource) <= $timestamp;
}
```

by this?

```php
public function isFresh($timestamp)
{
    return file_exists($this->resource) && filemtime($this->resource) <= $timestamp;
}
```

**Comment by javiereguiluz** [comment]
This PR just needs a final boost to finish it. Should I do the change mentioned in my previous comment? Thanks!

**Comment by jakzal** [comment]
@javiereguiluz I say go for it!

**Comment by javiereguiluz** [comment]
@jakzal OK. Changed. Thanks.

File: src/Symfony/Component/Config/Resource/FileResource.php [hunk 4]
```diff
@@ -56,11 +62,7 @@ public function getResource()
      */
     public function isFresh($timestamp)
     {
-        if (false === $this->resource || !file_exists($this->resource)) {
-            return false;
-        }
-
-        return filemtime($this->resource) <= $timestamp;
+        return file_exists($this->resource) && filemtime($this->resource) <= $timestamp;
```
Comment (fabpot) [inline src/Symfony/Component/Config/Resource/FileResource.php]: I think @Tobion meant: `return file_exists($this->resource) && @filemtime($this->resource) <= $timestamp;`.

Comment (javiereguiluz) [inline src/Symfony/Component/Config/Resource/FileResource.php]: OK. Changed.

**Comment by fabpot** [comment]
Thank you @javiereguiluz.

File: src/Symfony/Component/Config/Resource/FileResource.php [hunk 5]
```diff
@@ -56,11 +62,7 @@ public function getResource()
      */
     public function isFresh($timestamp)
     {
-        if (false === $this->resource || !file_exists($this->resource)) {
-            return false;
-        }
-
-        return filemtime($this->resource) <= $timestamp;
+        return file_exists($this->resource) && @filemtime($this->resource) <= $timestamp;
```
Comment (Tobion) [inline src/Symfony/Component/Config/Resource/FileResource.php]: `filemtime` returns false on error which means isFresh evalues to true in this case which is wrong.

File: src/Symfony/Component/Config/Resource/FileResource.php [hunk 6]
```diff
@@ -29,10 +29,16 @@ class FileResource implements SelfCheckingResourceInterface, \Serializable
...
      * @param string $resource The file path to the resource
+     *
+     * @throws \InvalidArgumentException
      */
     public function __construct($resource)
     {
         $this->resource = realpath($resource);
+
+        if (false === $this->resource) {
+            throw new \InvalidArgumentException(sprintf('The file "%s" does not exist.', $resource));
```
Comment (xabbuh) [inline src/Symfony/Component/Config/Resource/FileResource.php]: This change breaks our test suite. For example, the `ObjectRouteLoader` registers `FileResource` instances based on the filename returned by the `ReflectionClass`: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Routing/Loader/ObjectRouteLoader.php#L90 However, this fails when the object is a PHPUnit mock object, see https://travis-ci.org/symfony/symfony/jobs/106738857#L3760 (I am also not sure how this is working with classes stored in phar archives).

Comment (xabbuh) [inline src/Symfony/Component/Config/Resource/FileResource.php]: The question is how we fix this. I tend to fix the creation of `FileResource` instances to only build these objects when we handle "real" files. However, if we experience these issues in our test suite, isn't this something that will likely appear to other users too and (unexpectedly) break their applications?

Comment (stof) [inline src/Symfony/Component/Config/Resource/FileResource.php]: Well, for eval'd classes, the code was already weird before as it was creating a FileResource for something which is not a file. so I would consider it a bug in the ObjectRouteLoader

Comment (xabbuh) [inline src/Symfony/Component/Config/Resource/FileResource.php]: see #17672 and #17673

**Pull request closed without merging**